includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -1098,6 +1136,227 @@ private function sanitize_disallowed_attribute_values_in_node( DOMElement $node,
 		return $attributes_pending_removal;
 	}
 
+	/**
+	 * Validates the layout for the given tag. This involves checking the
+	 * layout, width, height, sizes attributes with AMP specific logic.
+	 *
+	 * @version 1911070201440
+	 * @link https://github.com/ampproject/amphtml/blob/1911070201440/validator/engine/validator.js#L3937
+	 *
+	 * Adapted from the <code>validateLayout</code> method found in validator.js of the
+	 * <a href="https://github.com/ampproject/amphtml">ampproject/amphtml</a> project.
+	 *
+	 * @param array[][]  $tag_spec Tag spec list.
+	 * @param DOMElement $node     Tag to validate.
+	 * @return bool True if layout is valid, false if validation fails.
+	 */
+	private function is_valid_layout( $tag_spec, $node ) {
+		// No need to validate if there are no specifications for the layout.
+		if ( ! isset( $tag_spec['amp_layout'] ) ) {
+			return true;
+		}
+
+		// Elements without a width nor height don't need to be validated.
+		if ( ! $node->hasAttribute( 'width' ) && ! $node->hasAttribute( 'height' ) ) {
+			return true;
+		}
+
+		$layout_attr = $node->getAttribute( 'layout' );
+		$allow_auto  = true;
+		$allow_fluid = $node->hasAttribute( 'layout' ) ? AMP_Rule_Spec::LAYOUT_FLUID === $layout_attr : true;
+
+		// The layout is not validated as yet to allow for the width and height to be done first.
+		$input_width = new AMP_CSS_Length( $node->getAttribute( 'width' ) );
+		$input_width->validate( $allow_auto, $allow_fluid );
+
+		if ( ! $input_width->is_valid() ) {
+			return false;
+		}
+
+		$input_height = new AMP_CSS_Length( $node->getAttribute( 'height' ) );
+		$input_height->validate( $allow_auto, $allow_fluid );
+
+		if ( ! $input_height->is_valid() ) {
+			return false;
+		}
+
+		// If no layout is supplied, there is nothing else to test.
+		if ( amp_is_attribute_empty( $layout_attr ) ) {
```

Comment:
Couldn't this do `$node->hasAttribute( 'layout' )`? 

Should the check not be done at the beginning of this method?

Reply:
Will be done at the beginning.